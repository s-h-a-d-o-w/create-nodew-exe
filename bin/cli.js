#!/usr/bin/env node
const argv = require('yargs')
.usage('Converts node.exe into a silent executable (no terminal will pop up) ' +
	'by changing only one byte.\n\n'+
	'Usage: $0 [options] <src> <dst>')
.demandCommand(2, 'You need to specify <src> and <dst> (or use --help for more)')
.help('h')
.alias('h', 'help')
.showHelpOnFail(false)
.argv;

console.log('Generating: ', argv._[1]);
require('../src/main')({
	src: argv._[0],
	dst: argv._[1],
});
console.log('SUCCESS');
