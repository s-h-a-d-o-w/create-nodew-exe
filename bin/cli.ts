#!/usr/bin/env node
import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { main } from "../src/main.js";

const argv = yargs(hideBin(process.argv))
  .usage(
    "Converts node.exe into a silent executable (no terminal will pop up) " +
      "by changing only one byte.\n\n" +
      "Usage: $0 [options] <src> <dst>",
  )
  .demandCommand(
    2,
    "You need to specify <src> and <dst> (or use --help for more)",
  )
  .help("h")
  .alias("h", "help")
  .showHelpOnFail(false)
  .parseSync();

console.log("Generating: ", argv._[1]);
main({
  src: String(argv._[0]),
  dst: String(argv._[1]),
});
console.log("SUCCESS");
