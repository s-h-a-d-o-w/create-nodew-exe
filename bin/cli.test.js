const fs = require('fs');
const crypto = require('crypto');

let spyLog, spyError;
beforeEach(() => {
	//spy = jest.spyOn(console, 'log');
	//jest.spyOn(console, 'error');

	// mockImplementation() to suppress all that CLI console output in the test logs
	// Output will still be visible in the snapshots because of .mock.calls
	spyLog = jest.spyOn(console, 'log').mockImplementation(() => {});
	spyError = jest.spyOn(console, 'error').mockImplementation(() => {});

	// process.exit() would exit jest as well but we still need to interrupt execution
	jest.spyOn(process, 'exit').mockImplementationOnce(
		() => { throw new Error('process.exit() was called.'); }
	);

	// Without resetting, yargs would remember args from previous tests
	jest.resetModules();

	// Purge previous console output before each test
	jest.clearAllMocks();
});


test('--help', () => {
	mockCLI(['--help'], 0);
});

test('no arguments', () => {
	mockCLI([], 1);
});

test('only src, no dst', () => {
	mockCLI(['src'], 1);
});

test('correct conversion', () => {
	process.argv = ['node', './bin/cli.js', './test/test_regular.exe', './test/test_silent.exe'];

	require('./cli.js');

	// Unless the source .exe changes, these check sums should obviously always stay the same
	expect(crypto.createHash('md5').update(fs.readFileSync('./test/test_regular.exe')).digest('hex'))
		.toEqual('037c3a6fcd6719e26d1bfc7a1a835a67');
	expect(crypto.createHash('md5').update(fs.readFileSync('./test/test_silent.exe')).digest('hex'))
		.toEqual('d1bf9c7e48771ed322827e5549716484');

	expect(spyLog.mock.calls).toMatchSnapshot('log');
	expect(spyError.mock.calls).toMatchSnapshot('error');

	fs.unlinkSync('./test/test_silent.exe');
});


const mockCLI = (args, expectedExitCode) => {
	process.argv = ['node', './bin/cli.js'].concat(args);

	expect(() => {
		require('./cli.js');
	}).toThrowError('process.exit() was called.');

	expect(spyLog.mock.calls).toMatchSnapshot('log');
	expect(spyError.mock.calls).toMatchSnapshot('error');
	expect(process.exit).toHaveBeenCalledWith(expectedExitCode);
};
