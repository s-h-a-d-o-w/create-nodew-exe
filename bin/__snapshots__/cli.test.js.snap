// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--help: error 1`] = `Array []`;

exports[`--help: log 1`] = `
Array [
  Array [
    "Converts node.exe into a silent executable (no terminal will pop up) by changing
only one byte.

Usage: cli.js [options] <src> <dst>

Options:
  --version   Show version number                                      [boolean]
  -h, --help  Show help                                                [boolean]",
  ],
]
`;

exports[`correct conversion: error 1`] = `Array []`;

exports[`correct conversion: log 1`] = `
Array [
  Array [
    "Generating: ",
    "./test/test_silent.exe",
  ],
  Array [
    "SUCCESS",
  ],
]
`;

exports[`no arguments: error 1`] = `
Array [
  Array [
    "You need to specify <src> and <dst> (or use --help for more)",
  ],
]
`;

exports[`no arguments: log 1`] = `Array []`;

exports[`only src, no dst: error 1`] = `
Array [
  Array [
    "You need to specify <src> and <dst> (or use --help for more)",
  ],
]
`;

exports[`only src, no dst: log 1`] = `Array []`;
